Telemetry/opt in welcome screen

Bamieh opened the pull request:
## Summary

Add Telemetry Opt in card to welcome screen. Closes https://github.com/elastic/dev/issues/1075

**The telemetry in opt in screen will be hidden in the following scenarios:**
1. OSS Kibana, since telemetry currently lives in x-pack.
2. User already chose to opt-in or opt-out.
3. Kibana config `xpack.telemetry.banner: false`.
4. Kibana config: `xpack.telemetry.enabled: false`.

**Scenarios:**
1. The user clicks "No":
	Opt in banner and this card will never be shown again. Telemetry is opted out.

2. The user clicks "Yes":
	Opt in banner and this card will never be shown again. Telemetry is opted in.

3. The user does not get the welcome screen
	Users who already have data on elasticsearch will not see the welcome screen, hence they will only see the opt in banner.


### Step 1:

![image](https://user-images.githubusercontent.com/6191849/63163364-df10eb00-c02d-11e9-980f-60c20a8b4d0e.png)


### Step 2:

![image](https://user-images.githubusercontent.com/6191849/63163583-77a76b00-c02e-11e9-8ad8-f5d3ac50f589.png)


### With example opened:

![image](https://user-images.githubusercontent.com/6191849/63163497-3a42dd80-c02e-11e9-97b9-a8ce9ba571ad.png)



### Dev Docs

Added Telemetry Opt in card to welcome screen. The aim is to increate increase opt in rate. Added tracking to measure the success of this change.

![image](https://user-images.githubusercontent.com/6191849/63163364-df10eb00-c02d-11e9-980f-60c20a8b4d0e.png)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9689/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9691/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9702/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9722/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9735/)

ryankeairns commented:
Quick look at the screenshots in the description, it reads that Step 1 has the first dot (under the box) as gray when it should be blue. In other words, the active slide should be associated with a blue dot. 

This may just be a mixup in the description, as I've not run this locally to confirm, but just pointing it out in the even it needs to change.

Looks good!

Bamieh commented:
@ryankeairns Thanks for the review! I just set the `disabled` attribute to true to the current card, so the grey one is the current one as it is not clickable. I would take any recommendation to improving the experience! Happy to chat over zoom/slack as well

ryankeairns commented:
@Bamieh for the active card, perhaps just use a regular `EuiIcon` (instead of EuiButtonIcon) and set the color to `euiColorPrimary`. For the other card (not currently visible), keep the `EuiButtonIcon` but set the color to `disabled` (don't actually set the button to `disabled`).

Unrelated, if the user clicks 'Try sample data' will they navigate away and potentially never reach the second card for telemetry?

Bamieh commented:
@ryankeairns 	I'll update the snapshots and ping you.

Users will be navigated to the opt in screen regardless what they click in the first card. After their opt in card selection they will be navigated to the sample data screen or home screen depending on what they selected in the first card.

cchaos commented:
What has been the conversation around the order of these boxes? They seem flipped to me.  You ask a question about where they want to go, but then there's a step in the middle that has nothing to do with the previous question. I would think you'd want to ask the telemetry question first so that on the second screen when you ask about how they want to get started, their answer is exactly where they go next.

<img src="https://d.pr/free/i/DfzYKK+" />

ryankeairns commented:
The order @cchaos suggests is the order we seemed to settle on at the end of the issue linked in the description.

ryankeairns commented:
Here's the scoop on the illustration (re: finding a higher res version).

@snide , et al, the blue/green/yellow illustration currently being used is from the old elastic.co website and has been redesigned for the new site (see below). That said, it currently is designed to work on a blue background.

I chatted briefly with Jess and we're wondering if we can use a blue background here... perhaps we can just use a blue background on both cards, for that matter. Otherwise, Creative Services would need some extra time to recolor the illustration to work on a white background.

#### Preview: How it looks on the new website
<img width="601" alt="Screenshot 2019-07-29 15 28 00" src="https://user-images.githubusercontent.com/446285/62081711-cb394c80-b218-11e9-83b9-5382c9b2627f.png">

#### Transparent PNG (used on the website; from the preview above)
![illustration-home-hero-get-started-elasticsearch-585x530](https://user-images.githubusercontent.com/446285/62081675-bb216d00-b218-11e9-924a-c53858b41a07.png)

cchaos commented:
A. I love the new artwork
B. I'd love to see more color. +1 to the blue background for both.
C. I'm thinking the artwork for the Let's Get Started panel might need to be updated to reflect the new nav style. But that can be a separate PR.

bmcconaghy commented:
@snide Definitely can't be opt out (GDPR).

alexfrancoeur commented:
Pretty much all of my feedback has already been addressed by @cchaos @snide & @ryankeairns, but I'd still like to add some additional thoughts.

I agree that the order feels backward right now. But I also feel like it's strange to immediately prompt a user with a telemetry. This is definitely a UI/UX call but would it make sense to have a welcome to Kibana card that is just generic? Maybe a one or two liner borrowing some similar messaging from our guides page?

> Welcome
> Welcome to Kibana! Your one stop shop for data visualization & exploration, management of your Elastic Stack and home for Elastic solutions! (_obviously, really need Gails help here_)

Welcome to Kibana `[Next]` ➡️  Help us improve the Elastic Stack `[Yes]` `[No]` ➡️ Let's get started `[Try our sample data]` `[Explore on my own]`

Maybe this is too many cards and unnecessary, but it doesn't throw telemetry right in your face as the absolute first thing you see. Thoughts from the group?

snide commented:
> I agree that the order feels backward right now. But I also feel like it's strange to immediately prompt a user with a telemetry

We do this now pretty much. Just in banner form.

> Welcome [Next] ➡️ Help us improve the Elastic Stack [Yes] [No] ➡️ Let's get started [Try our sample data] [Explore on my own]

I don't think we need a roadblock on a new install of Kibana for a paragraph on what Kibana is (hopefully our marketing / cloud efforts did a decent job telling them that arleady). I'd keep the pain of the user as minimal as possible. The least amount of clicks as possible. PMs and ELT have told us telemetry is really important, so this is a good method to make action and possibly get the numbers up.

Remember too that the new homepage won't need the sample data notice, since that's built into that experience. When that comes, the flow would be:

Help us improve the Elastic Stack [Yes] [No] ➡️ Homepage with add data as the core focus + a bunch of learn about kibana stuff

Whether you want to wait for that is up to you all, but in the meantime I'd suggest @cchaos' flow and just get this out of the users way as fast a possible so they can get to adding data.

But yes. We need that new homepage. It helps this early flow tremendously.

cchaos commented:
I also would like to point out, though I did not look through this code to see if the logic has changed, that this screen (the full takeover) only gets presented to the user if they have no data. This means if they already have data in ES, they'll never see this telemetry card.

That said, if they **do** see this card and answer the question, then they should never see the current telemetry (blue) banner.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12773/)

Bamieh commented:
I've updated the design according to feedback. i've removed the 2 dots at the bottom of the cards. I like the experience without it more, I can add it back if you see otherwise. @snide @cchaos @ryankeairns

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12788/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12798/)

alexfrancoeur commented:
Woah! This looks fantastic @Bamieh. One quick comment. I noticed the banner was still there when I loaded the screen. I think I have the latest pr (just pulled down, `yarn kbn clean`, `yarn kbn bootstrap`) and cleared local storage. Is anyone else still seeing the banner?

snide commented on src/legacy/core_plugins/kibana/public/home/components/telemetry_opt_in/telemetry_opt_in_card.tsx:
```diff
@@ -0,0 +1,81 @@
+              defaultMessage="Yes"
+            />
+          </EuiButton>
+          <EuiButton
```
Change this one to `EuiButtonEmpty`. It will highlight the Yes answer more.

alexfrancoeur replied:
I thought about this as well, but I _think_ we need to keep them colored equally for GDPR reasons. I believe that's why we emphasize them equally in the banner today where that wasn't always the case

Bamieh replied:
We have to have both buttons exactly the same to not introduce bias towards clicking yes according to our legal team (I believe it has to do with GDPR). Similar to what we have in the banner.

snide replied:
Fair enough. Makes sense.

alexfrancoeur commented:
I'm running into weird issues I think. When loading with no data and cleared cache incognito, I'll get the banner without the welcome screen. If I refresh, I'm then presented with the welcome screen. I can consistently reproduce this. This is likely related to the race condition you mentioned @Bamieh

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12907/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13041/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13048/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13052/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13055/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13119/)

alexfrancoeur commented:
If I hit no for the telemetry request, I no longer see the banner. However, it is there for a quick second as Kibana loads. I realize that showing the quick flash of the Kibana before the splash screen is a bug that has been there for a little while, but is there anything we can do to avoid showing the banner? Is anyone else seeing this?

![Aug-19-2019 10-10-39](https://user-images.githubusercontent.com/25181823/63272310-d4429a00-c269-11e9-90c6-ab714e85fd88.gif)

Bamieh commented:
@alexfrancoeur This (bug) is not caused by this PR. The banner is loaded as soon as the telemetry code is executed. I can add a quick fix by setting a timeout so that it gets populated after the loading screen is covering the screen.

I do prefer However to wait until we move telemetry to OSS then i'd be able to trigger the banner only if the welcome screen is not going to show. Currently I cannot do this easily because the logic for checking if we want to show the welcome screen lives in OSS while the telemetry banner in x-pack.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13579/)

tsullivan commented on src/legacy/core_plugins/kibana/public/home/components/home.js:
```diff
@@ -247,6 +251,10 @@ export class Home extends Component {
+  fetchTelemetry: PropTypes.func.isRequired,
+  getTelemetryBannerId: PropTypes.func.isRequired,
+  setOptIn: PropTypes.func.isRequired,
+  showTelemetryOptIn: PropTypes.bool.isRequired,
```
it was a little hard to determine that this variable is a boolean, without reading down to here. Maybe `shouldShowTelemetryOptIn`?

tsullivan commented on src/legacy/core_plugins/kibana/public/home/components/home.test.js:
```diff
@@ -17,30 +17,14 @@
  * under the License.
  */
 
+import './home.test.mocks';
```
👍

tsullivan commented on src/legacy/core_plugins/kibana/public/home/kibana_services.js:
```diff
@@ -18,9 +18,23 @@
+import { TelemetryOptInProvider } from './telemetry_opt_in';
 
 export let indexPatternService;
+export let showTelemetryOptIn;
```
I mentioned it before, but I'd really like to see a better boolean naming for this one. I realize it's mostly existing code, but this doesn't seem a very typical way to for a service to expose a value.

tsullivan commented on x-pack/legacy/plugins/telemetry/public/services/telemetry_opt_in.ts:
```diff
@@ -0,0 +1,8 @@
+ */
+
+// @ts-ignore
+export { TelemetryOptInProvider } from '../../../../../../src/legacy/core_plugins/kibana/public/home/telemetry_opt_in'; // eslint-disable-line
```
`// eslint-disable-line` is not necessary here

This include needs to be tested out with build Kibana too, since in built Kibana, the location of `x-pack` moves. Relative imports out of x-pack this way, do not work afaik

Bamieh replied:
I'm disabling eslint as it spreads the `export` statement into 3 lines as the import string is long. Then the `@ts-ignore` does not ignore the right line. Moving the `@ts-ignore` line between those 3 lines of the `export` statement also makes the linter complain.

Bamieh replied:
I did test this in a production build and it works fine; there are places in x-pack where its requiring code from src as well in a 100 different files (searched x-pack folder for `../src` excluding test files)

tsullivan commented on x-pack/legacy/plugins/telemetry/public/services/telemetry_opt_in.test.mocks.js:
```diff
@@ -0,0 +1,27 @@
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { injectedMetadataServiceMock } from '../../../../../../src/core/public/mocks';
```
Regarding my earlier comment about relative imports going out of OSS into X-Pack, that might not apply to test code

tsullivan commented:
@Bamieh 
> The user does not get the welcome screen

How do we test that in this PR branch?

> I can add a quick fix by setting a timeout so that it gets populated after the loading screen is covering the screen.

I am ++ for a quick fix of the flash of the upper banner. Would it be possible to tweak some timing around there?

Bamieh commented:
@tsullivan thanks for the review. To test this scenario:

- The user does not get the welcome screen
Users who already have data on elasticsearch will not see the welcome screen, hence they will only see the opt in banner.


You can press `esc` to skip the welcome screen dialog without specifying opt in or opt out. You should see the opt in banner then.

Another way to test this is by having some data in ES, then the welcome screen  will not show in the first place; leaving the users with the opt in banner to enable telemetry.

Bamieh commented:
@tsullivan @alexfrancoeur For the banner timeout enhancement I will create a separate PR for it as it seems out of scope for this PR.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14267/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/715/)
